### PR TITLE
[JENKINS-31582] Using ArgumentsAction.getResolvedArguments to better handle JENKINS-45109

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/workflow-api-plugin/pull/44.

@reviewbybees

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java [hunk 1]
```diff
@@ -394,7 +395,7 @@ public void testUnusualStepInstantiations() throws Exception {
                 " node('master') { \n" +
                 "   writeFile text: 'hello world', file: 'msg.out'\n" +
                 "   step([$class: 'ArtifactArchiver', artifacts: 'msg.out', fingerprint: false])\n "+
-                "   withEnv(['CUSTOM=val']) {\n"+  //Symbol-based, because withEnv is a metastep
+                "   withEnv(['CUSTOM=val']) {\n"+  //Symbol-based, because withEnv is a metastep; TODO huh? no it is not
```
Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java]: That i don't 😖 .

File: src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java [hunk 2]
```diff
@@ -423,4 +424,25 @@ public void testUnusualStepInstantiations() throws Exception {
         Assert.assertEquals("CUSTOM=val", (String)((ArrayList) ob).get(0));
         testDeserialize(run.getExecution());
     }
+
+    @Test
+    public void testReallyUnusualStepInstantiations() throws Exception {
```
Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java]: Ok, I really like this test name. =)

File: src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java [hunk 3]
```diff
@@ -254,7 +255,13 @@ Object sanitizeObjectAndRecordMutation(@CheckForNull Object o, @CheckForNull Env
         if (modded != tempVal) {
             // Sanitization stripped out some values, so we need to record that and return modified version
             this.isUnmodifiedBySanitization = false;
-            return modded;
+            if (o instanceof UninstantiatedDescribable) {
+                // Need to retain the symbol.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java]: This is a fix to what is _stored_. Without this, even a fleshed-out version of `getResolvedArguments` would have no hope, because previously the storage would be something like

    delegate → [<anonymous> → 'TEST-*.xml']

with no record being kept that the original delegate was in fact `junit` ~ `JUnitResultArchiver`.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java (hunk 1 above)
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java]: Not sure what @svanoort thought he was testing in this case. If the idea was to test a symbol-based block-scoped metastep (like `configFileProvider`), it is not doing that.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java [hunk 4]
```diff
@@ -423,4 +424,25 @@ public void testUnusualStepInstantiations() throws Exception {
...
+        WorkflowJob job = r.jenkins.createProject(WorkflowJob.class, "unusualInstantiation");
+        job.setDefinition(new CpsFlowDefinition(
+                " node() {\n" +
+                "   writeFile text: 'hello world', file: 'msg.out'\n" +
+                "   step(new hudson.tasks.ArtifactArchiver('msg.out'))\n" + // note, not whitelisted
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java]: Seems that this case was implicitly handled in the code but not tested.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/nodes/StepNodeTest.java [hunk 5]
```diff
@@ -78,4 +83,87 @@
...
+            "node {\n" +
+            "  wrap([$class: 'ConfigFileBuildWrapper', managedFiles: []]) {\n" +
+            "    writeFile text: '''<testsuite name='a'><testcase name='c'><error>failed</error></testcase></testsuite>''', file: 'x.xml'\n" +
+            "    step([$class: 'JUnitResultArchiver', testResults: 'x.xml'])\n" +
+            "  }\n" +
+            "}", true));
+        WorkflowRun b = r.assertBuildStatus(Result.UNSTABLE, p.scheduleBuild2(0));
+        List<FlowNode> coreStepNodes = new DepthFirstScanner().filteredNodes(b.getExecution(), new NodeStepTypePredicate("step"));
+        assertThat(coreStepNodes, hasSize(1));
+        assertEquals("junit", coreStepNodes.get(0).getDisplayFunctionName());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/nodes/StepNodeTest.java]: We may as well show the nice view even if you have not yet updated your script to take advantage of simplified syntax.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/nodes/StepNodeTest.java [hunk 6]
```diff
@@ -78,4 +83,87 @@
...
+            "node {\n" +
+            "  wrap(new org.jenkinsci.plugins.configfiles.buildwrapper.ConfigFileBuildWrapper([])) {\n" +
+            "    writeFile text: '''<testsuite name='a'><testcase name='c'><error>failed</error></testcase></testsuite>''', file: 'x.xml'\n" +
+            "    step(new hudson.tasks.junit.JUnitResultArchiver('x.xml'))\n" +
+            "  }\n" +
+            "}", false));
+        WorkflowRun b = r.assertBuildStatus(Result.UNSTABLE, p.scheduleBuild2(0));
+        List<FlowNode> coreStepNodes = new DepthFirstScanner().filteredNodes(b.getExecution(), new NodeStepTypePredicate("step"));
+        assertThat(coreStepNodes, hasSize(1));
+        assertEquals("junit", coreStepNodes.get(0).getDisplayFunctionName());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/nodes/StepNodeTest.java]: A corner case but we get it for free anyway.
